### ISSUE TITLE
Custom SynchronizationContext is removed by v3

### ISSUE THREAD

**Issue by jbe2277** [xunit/xunit#3014]
xUnit replaces any existing SynchronizationContext during test execution.
- This will happen for `public void` test methods as well without using `async` at all.

xUnit behaves here different depending on the version:
- 2.7.0: Does not replace the SynchronizationContext; works as expected ✅
- 2.9.0: Replace the SynchronizationContext with the `Xunit.Sdk.AsyncTestSyncContext`
- v3 (0.2.0-pre.69): Replace the SynchronizationContext with `null`

## Reproduction
1. Create a xUnit project with Visual Studio 2022
2. Add NuGet package [System.Waf.UnitTesting.Core](https://www.nuget.org/packages/System.Waf.UnitTesting.Core)
3. Replace UnitTest1.cs code with:

```C#
// NuGet: System.Waf.UnitTesting.Core

namespace TestProject1;

public class UnitTest1 : IDisposable
{
    public UnitTest1()
    {
        Context = System.Waf.UnitTesting.UnitTestSynchronizationContext.Create();
    }

    public System.Waf.UnitTesting.UnitTestSynchronizationContext Context { get; }

    public void Dispose() => Context.Dispose();
    
    [Fact]
    public void Test1()
    {
        Assert.Equal(Context, SynchronizationContext.Current);
    }
}
```

4. Run the test with different xUnit versions to see the different behavior. Only version 2.7.0 will pass.

## Motivation
We use the `UnitTestSynchronizationContext` to simulate a more realistic async behavior of UI applications (e.g. MAUI, Xamarin, WinUI3, WPF, Windows Forms). Those application models come with a SynchronizationContext that allows to schedule work in the UI thread. This is the default for `await` that the code after this statement runs again within the UI thread.

**Example:** This Assert fails because the statement after `await` is run in another thread. In UI applications the assert would pass.

```C#
public class UnitTest2
{
    [Fact]
    public async Task Test2()
    {
        var mainThread = Thread.CurrentThread;

        await Task.Delay(10);

        // Verify that after "await" we are back in the main thread -> that would be the case for the main UI thread (Dispatcher).
        Assert.Equal(mainThread.ManagedThreadId, Thread.CurrentThread.ManagedThreadId);
    }
}
```

**Comment by bradwilson** [comment]
## v2

Setting a synchronization context (or more than one) has long been the norm for us. `AsyncTestSyncContext` goes all the way back into v1.

Here are some more recent versions:

v2 2.6.0: Replace the SynchronizationContext with the `AsyncTestSyncContext`
v2 2.6.1: Replace the SynchronizationContext with the `AsyncTestSyncContext`
v2 2.6.2: Replace the SynchronizationContext with the `AsyncTestSyncContext`
v2 2.6.3: Replace the SynchronizationContext with the `AsyncTestSyncContext`
v2 2.6.4: Replace the SynchronizationContext with the `AsyncTestSyncContext`
v2 2.6.5: Replace the SynchronizationContext with the `AsyncTestSyncContext`
v2 2.6.6: Replace the SynchronizationContext with the `AsyncTestSyncContext`
v2 2.7.0: Replace the SynchronizationContext with the `MaxConcurrencySyncContext`
v2 2.7.1: Replace the SynchronizationContext with the `MaxConcurrencySyncContext`
v2 2.8.0: Replace the SynchronizationContext with the `AsyncTestSyncContext`
v2 2.8.1: Replace the SynchronizationContext with the `AsyncTestSyncContext`
v2 2.9.0: Replace the SynchronizationContext with the `AsyncTestSyncContext`

I don't see the same thing you see, because you presumably left out a step in your repro: you've turned off the need for `MaxConcurrencySyncContext` by setting unlimited parallel threads:

https://github.com/xunit/xunit/blob/6b86235b1b2fa3bda92be6dde1cca389d72f1278/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestAssemblyRunner.cs#L79-L92

That means, in 2.7.0 and 2.7.1, you would see us not putting in place a sync context. What's different in 2.7.0/2.7.1 than everything else? From the [2.7.0 release notes](https://xunit.net/releases/v2/2.7.0):

![image](https://github.com/user-attachments/assets/76e8906e-eb57-4008-9c7b-4694e93e14ce)

And the [2.8.0 release notes](https://xunit.net/releases/v2/2.8.0):

![image](https://github.com/user-attachments/assets/f48d3a2a-0b81-4165-a6bc-f763d3656251)

That said, even when `AsyncTestSyncContext` is what you see, it delegates to the sync context that was there before it. The entire purpose of the context is to wrap calls to `Post` to increment and decrement counters as a way of measuring `async void` operations in flight. We always delegate the actual work to the context that was there before us:

https://github.com/xunit/xunit/blob/6b86235b1b2fa3bda92be6dde1cca389d72f1278/src/xunit.execution/Sdk/AsyncTestSyncContext.cs#L42-L104

I replaced your test with this:

```csharp
using System;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace TestProject1;

public class MyCustomSyncContext : SynchronizationContext, IDisposable
{
    public void Dispose() { }

    public override void Post(SendOrPostCallback d, object? state)
    {
        base.Post(d, state);
    }

    public override void Send(SendOrPostCallback d, object? state)
    {
        base.Send(d, state);
    }
}

public class UnitTest1 : IDisposable
{
    public UnitTest1()
    {
        Context = new MyCustomSyncContext();

        SynchronizationContext.SetSynchronizationContext(Context);
    }

    public MyCustomSyncContext Context { get; }

    public void Dispose() => Context.Dispose();

    [Fact]
    public async Task Test1()
    {
        await Task.Delay(10);

        Assert.Equal(Context, SynchronizationContext.Current);
    }
}
```

I placed breakpoints on the Post and Send methods of `MyCustomSyncContext`, and breakpoints on the `Task.Delay()` line and `Assert.Equal` line of the test. First it hits the `Task.Delay`, then it hits the `Post` method of my sync context, then it hits the `Assert.Equal`, then it hits `Post` several more times as the test cleans up.

This is the behavior I would expect, and it means any sync context you set in the constructor should be honored inside the test method, **_even though the `SynchronizationContext.Current` does not point to your sync context_**, because we are delegating to yours. You can even stop on the break point in the `Task.Delay`, look at `SynchronizationContext.Current` and dive into the non-public members and see that the `innerContext` is pointing to `MyCustomSyncContext`:

![image](https://github.com/user-attachments/assets/e657e67b-a580-4ef2-a6fc-2bb9d076f425)

v2 is working as expected here.

## v3

The fact that the sync context here is `null` appears to be a bug.

**Comment by jbe2277** [comment]
Thanks for the detailed information.

**Comment 1:**

> I don't see the same thing you see, because you presumably left out a step in your repro: you've turned off the need for MaxConcurrencySyncContext by setting unlimited parallel threads:

I created again a new xUnit (.NET 8) project with Visual Studio 17.10.4 and added the code for reproduction. I have updated all NuGet packages to the latest version, only xunit is set to version 2.7.0. I have not changed any setting regarding unit test execution (e.g. unlimited parallel threads). The test passes when I run it from Visual Studio Test Explorer or via CLI `dotnet test`.

**Comment 2:**

Our unit tests for production code (MAUI app) work fine with xUnit 2.7.0. After updating it to >= 2.8.0 we run into deadlocks when running multiple unit tests. Probably the delegation done by `AsyncTestSyncContext` to our `UnitTestSynchronizationContext` is causing the deadlocks.

**Comment by jbe2277** [comment]
Reproduction code of the deadlock we see with xUnit 2.9.0:

```C#
// NuGet: System.Waf.UnitTesting.Core
public class UnitTestDeadlock : IDisposable
{
    public UnitTestDeadlock()
    {
        Context = System.Waf.UnitTesting.UnitTestSynchronizationContext.Create();
    }

    public System.Waf.UnitTesting.UnitTestSynchronizationContext Context { get; }

    public void Dispose() => Context.Dispose();

    [Fact]
    public void TestDeadlock()
    {
        DoBackgroundWork();  // do not await here
    }

    private async void DoBackgroundWork()
    {
        await Task.Delay(50);
    }
}
```

If the `UnitTestSynchronizationContext` is created within the test method it works correct (no deadlock occurs):

```C#
// NuGet: System.Waf.UnitTesting.Core
public class UnitTestWorkaround()
{
    [Fact]
    public void TestWorkaround()
    {
        using var context = System.Waf.UnitTesting.UnitTestSynchronizationContext.Create();        
        DoBackgroundWork();  // do not await here
    }

    private async void DoBackgroundWork()
    {
        await Task.Delay(50);
    }
}
```

**Comment by bradwilson** [comment]
> If the `UnitTestSynchronizationContext` is created within the test method it works correct (no deadlock occurs)

Yes, looking at the source for `UnitTestSynchronizationContext`, the design of this class does not support being delegated to. It's basically waiting for the call to Dispose to actually finish running the test code, which is well outside the time when such code should be running. It fails to delegate properly because the async test sync context has detected your usage of `async void` and is waiting for that code to finish running... which it will never do, because the `UnitTestSynchronizationContext` has not run it yet (and never will, because of this deadlock).

The design of this class does not support being used in any way other than inside the test method. The fact that it can occasionally work outside the test method is purely coincidence and luck.

**Comment by bradwilson** [comment]
FWIW, with v3 when the sync context preservation bug is fixed, you will still be able to use your sync context in the ctor (assuming you don't [enable `aggressive` parallelism](https://xunit.net/docs/configuration-files#parallelAlgorithm), and assuming you don't ever introduce a third party extension that sets a sync context), even though it's still contractually violating the notion of when test code should be run. We continue to monitor `Dispose`/`DisposeAsync` for exceptions and will report failures (not because we expect there to generally be test code in `Dispose`, but because failures here should become test failures regardless).

**Comment by bradwilson** [comment]
v3 fix available in `0.3.0-pre.17` https://xunit.net/docs/using-ci-builds
